### PR TITLE
feat: make module window check more robust

### PR DESCRIPTION
**Pull request by cmackenthun**
We started using this rule in one of our repos (and are loving it so far!) and we noticed it was marking files that have patterns like this as needing the client directive 

```
const HREF = typeof window === 'undefined' ? undefined : window.location.href;
```

This is fine to execute on the server or the client because we are checking to make sure the window exists before accessing it. 

So I took a stab at updating the lint rule to account for this case. This is my first time dabbling in writing a lint rule, so please feel free to make suggestions! Overall we just want to get this case accounted for so we don't run into more issues with files being improperly marked 🙂

### CODE REVIEW

File: src/rules/use-client.ts [hunk 1]
```diff
@@ -208,11 +208,40 @@ const create = Components.detect(
...
+
+        // check if the window usage is behind a typeof window === 'undefined' check
+        const conditionalExpressionNode = node.parent?.parent;
+        const isWindowCheck =
+          conditionalExpressionNode?.type === "ConditionalExpression" &&
+          conditionalExpressionNode.test?.type === "BinaryExpression" &&
+          conditionalExpressionNode.test.left?.type === "UnaryExpression" &&
+          conditionalExpressionNode.test.left.operator === "typeof" &&
+          conditionalExpressionNode.test.left.argument?.type === "Identifier" &&
+          conditionalExpressionNode.test.left.argument?.name === "window" &&
```
Comment (roginfarrer) [inline src/rules/use-client.ts]: Does this catch other places where we might make this check? Like in an `if` statement:

```js
if (typeof window !== 'undefined') {
}
```

We could use `Identifier` as the node entrypoint, and checking whether it's `'window'`. Then navigate the tree to see if it's used safely. Should this also work with other browser globals?

I think for legibility I might recommend extracting this logic into a "checkIfSafeBrowserGlobalUse" function outside of this function (where it takes `node` as an argument).

File: src/rules/use-client.ts [hunk 2]
```diff
@@ -208,11 +208,40 @@ const create = Components.detect(
...
+          conditionalExpressionNode.test?.type === "BinaryExpression" &&
+          conditionalExpressionNode.test.operator === "!==";
+
+        // checks to see if window is being accessed safely behind a window check
+        const isSafelyBehindWindowCheck =
+          (isWindowCheck &&
+            !isNegatedWindowCheck &&
+            conditionalExpressionNode.alternate === node?.parent) ||
+          (isNegatedWindowCheck &&
+            conditionalExpressionNode.consequent === node?.parent);
```
Comment (roginfarrer) [inline src/rules/use-client.ts]: What is `conditionalExpressionNode.consequent === node?.parent` meant to check? Not sure what this is for

File: src/rules/use-client.test.ts [hunk 3]
```diff
@@ -99,6 +106,13 @@ function Bar() {
   window.addEventListener('scroll', () => {})
   return <div />;
 }`,
+        },
+        {
+          code: `const HREF = typeof window === 'undefined' ? window.location.href : window.location.href.slice(0,10);`,
```
Comment (roginfarrer) [inline src/rules/use-client.test.ts:152]: Good edge case to catch!

File: .changeset/honest-peas-peel.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+"eslint-plugin-react-server-components": patch
```
Comment (roginfarrer) [inline .changeset/honest-peas-peel.md]: This is kind of a new feature, so I think it warrants a minor

```suggestion
"eslint-plugin-react-server-components": minor
```

**Review by roginfarrer — COMMENTED** [review]
Nice, thanks for creating a PR!

File: src/rules/use-client.ts (hunk 1 above)
Comment (cmackenthun) [reply src/rules/use-client.ts]: Yeah I'm not sure about the if statement or not, but I don't think this would cover that. I was mostly going for the simplest solution to the one edge case we were seeing. I can look into using the `Identifier` as an entry point but am still pretty new to writing a lint rule, so that might take a bit more effort for me. Would we still need some kind of check under the `MemberExpression` entrypoint to make sure we skip erroring on these safe checks?

Agreed on the legibility part 👍 I can do that for sure.

File: src/rules/use-client.ts (hunk 2 above)
Comment (cmackenthun) [reply src/rules/use-client.ts]: This is basically to cover either `typeof window === undefined ? <server code> : <client code>` OR `typeof window !== undefined ? <client code> : <server code>`. Basically just checking if the alternate or the consequent is the client code based on the `isNegatedWindowCheck` value

File: src/rules/use-client.ts (hunk 1 above)
Comment (cmackenthun) [reply src/rules/use-client.ts]: Oh nvm, I think you're advocating for removing the MemberExpression entrypoint in favor of a starting at the Identifier level. I'll investigate that a bit 👍

File: src/rules/use-client.test.ts [hunk 5]
```diff
@@ -24,29 +24,21 @@ describe("use client", () => {
...
-    export function Foo() {
-      return <div />;
-    }
-    export function useTheme() {
-      const context = React.useContext(context);
-      React.useEffect(() => {
-        window.setTimeout(() => {});
-      });
-      return context;
-    }`,
```
Comment (cmackenthun) [inline src/rules/use-client.test.ts:49]: I moved these to the invalid section cause honestly I'm not sure why they were valid to being with. From my local testing if I have `const context = React.createContext()` at the top of a component file and use that component in a server context I get a runtime error. If there's something I'm missing please let me know!

File: src/rules/use-client.test.ts [hunk 6]
```diff
@@ -24,29 +24,21 @@ describe("use client", () => {
...
+          code: `const HREF = typeof window !== 'undefined' ? window.location.href : '';`,
+        },
+        {
+          code: `const el = typeof document === 'undefined' ? undefined : document.createElement('element');`,
+        },
+        {
+          code: `const foo = "bar";
+function Bar() {
+  if(typeof document !== 'undefined')
```
Comment (cmackenthun) [inline src/rules/use-client.test.ts:38]: Added functionality for document checks too since it was easy enough to do so 👍

File: src/rules/use-client.test.ts (hunk 5 above)
Comment (cmackenthun) [reply src/rules/use-client.test.ts:49]: If we do move forward with it and you feel like this warrants a major LMK 👍

File: src/rules/use-client.ts [hunk 7]
```diff
@@ -26,6 +27,8 @@ const browserOnlyGlobals = Object.keys(globals.browser).reduce<
   return acc;
 }, new Set());
 
+const validGlobalsForServerChecks = new Set(["document", "window"]);
```
Comment (roginfarrer) [inline src/rules/use-client.ts]: Wouldn't this be true for any of the browser globals? `typeof HTMLElement !== 'undefined'`

File: src/rules/use-client.ts [hunk 8]
```diff
@@ -113,12 +116,97 @@ const create = Components.detect(
       });
     }
 
+    function findFirstParentOfType(
```
Comment (roginfarrer) [inline src/rules/use-client.ts]: These pure functions can be moved to the bottom of the file, outside of the program

File: src/rules/use-client.ts [hunk 9]
```diff
@@ -113,12 +116,97 @@ const create = Components.detect(
...
+
+      let serverBranch = null;
+
+      if (!isGlobalClientPropertyCheck) {
+        return { isGlobalClientPropertyCheck, serverBranch };
+      }
+
+      //@ts-expect-error
+      const { parent } = node;
+      if (!parent) {
```
Comment (roginfarrer) [inline src/rules/use-client.ts:139]: Does this fix the error?
```suggestion
      const parent = node?.parent
      if (!parent) {
```

File: src/rules/use-client.test.ts (hunk 5 above)
Comment (roginfarrer) [reply src/rules/use-client.test.ts:49]: I think it would probably be best as a major. Happy to leave this in and cut the next release as a major

Comment (cmackenthun) [reply src/rules/use-client.test.ts:49]: @roginfarrer When you say "leave this in" are you saying I should leave my changeset as is? Or update that to a major? Or just remove it all together?

File: src/rules/use-client.ts (hunk 7 above)
Comment (cmackenthun) [reply src/rules/use-client.ts]: Yeah should be 👍 I only have ever seen these two checks being used, but I can update to use `browserOnlyGlobals` 👍

File: src/rules/use-client.ts (hunk 9 above)
Comment (cmackenthun) [reply src/rules/use-client.ts:139]: No, the error comes from the `parent` field apparently not being a part of the `BinaryExpression` type 🤔 

`Property 'parent' does not exist on type 'BinaryExpression'.`

File: src/rules/use-client.test.ts (hunk 5 above)
Comment (cmackenthun) [reply src/rules/use-client.test.ts:49]: @roginfarrer just wondering if you have a chance to answer ☝️ and then I can make the changes and get this merged

Comment (roginfarrer) [reply src/rules/use-client.test.ts:49]: Do you know why they pass in both valid and invalid sections? Or did your changes make them invalid?

Comment (cmackenthun) [reply src/rules/use-client.test.ts:49]: Yeah my changes made them invalid. And when I tested the scenario in a small next project, I would get runtime client only errors in the dev server related to the React Context so I think having them be invalid is correct. Please feel free to double check that though!
